### PR TITLE
Fix MSVC LNK1170

### LINKED ISSUE #12329: Meson generates potentially invalid response files for MSVC's Linker (link.exe)
**Describe the bug**

When linking an absurd amount of object files (in my case around 4,000) a response file gets made, which is expected.

This file has one single line, which ends up invoking error LNK1170 in the MSVC Linker because the entire line is longer than 128KB. 

By replacing one variable in the build.ninja file post-configure this will work.

**Workaround**

Letting it configure, and changing

```
rule cpp_LINKER_RSP
 command = "link" @$out.rsp
 rspfile = $out.rsp
 rspfile_content = $ARGS /MACHINE:x64 /OUT:$out $in $LINK_ARGS
 description = Linking target $out
```

to this:

```
rule cpp_LINKER_RSP
 command = "link" @$out.rsp
 rspfile = $out.rsp
 rspfile_content = $ARGS /MACHINE:x64 /OUT:$out $in_newline $LINK_ARGS
 description = Linking target $out
 ```

in the build.ninja file. The build will work correctly (and is also what the [ninja docs say](https://ninja-build.org/manual.html))

>  in_newline: the same as $in except that multiple inputs are separated by newlines rather than spaces. (For use with $rspfile_content; this works around a bug in the MSVC linker where it uses a fixed-size buffer for processing input.) 

TL;DR change $in to $in_newline for _RSP rules when rsp file syntax is RSPFileSyntax.MSVC.

I initially considered making a pull request for this but I am not experienced enough with both python and the meson code-base to feel comfortable enough making this change without it looking like found VHS footage.

**To Reproduce**

- Have a project with a lot of paths or a few very long paths resulting in a single line .rsp file over 128KB
- Build and link a shared library using MSVC as the target toolchain.

**Expected behavior**

Shared library gets made.

**system parameters**
* no cross compiling.
* Windows 11 
* Python 3.11.6
* Meson 1.2.1
* Ninja 1.11.1

### PR DESCRIPTION
**Pull request by yretenai**
This fixes #12329 by replacing `$in` with `$in_newline` when `rspfile quote style` is `RSPFileSyntax.MSVC`

### CODE REVIEW

File: mesonbuild/backend/ninjabackend.py [hunk 1]
```diff
@@ -237,8 +237,10 @@ def _quoter(x, qf = quote_func):
         return ninja_quote(qf(str(x)))
 
     def write(self, outfile: T.TextIO) -> None:
+        rspfile_args = self.args
         if self.rspfile_quote_style is RSPFileSyntax.MSVC:
             rspfile_quote_func = cmd_quote
+            rspfile_args = [NinjaCommandArg('$in_newline', arg.quoting) if arg.s == '$in' else arg for arg in rspfile_args]
```
Comment (tristan957) [inline mesonbuild/backend/ninjabackend.py:243]: Why not just replace it where the `$in` is generated?

Comment (yretenai) [reply mesonbuild/backend/ninjabackend.py:243]: Because as far as I can tell the RSP rule is a duplicate of the regular rule, the generator that adds `$in` is in a state unaware of whether or not RSP is used. 

There would be a less than insignificant refactor to the rule writer because the lines https://github.com/mesonbuild/meson/blob/523a27c6f0b40c7ed1cc04d4d5c2d74cc9f6b30e/mesonbuild/backend/ninjabackend.py#L245-L251 assume the command is identical for both RSP and regular rules. 

In other words, everything related to the RSP generation would have to be duplicated into every generator that calls `self._rsp_options(compiler)` right before calling `self.add_rule(NinjaRule(...` and the writer would have to be refactored.

**Comment by nirbheek** [comment]
Could you please update the commit message to have an explanation of what it's fixing, exactly? For example:

```
Fix MSVC linker error LNK1170 when using an rspfile

The MSVC linker errors out if a line in the rspfile is too long, so use the 
pre-existing ninja feature $in_newline to split inputs into separate lines.
```
